core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PrioritizeUtilizationExecutionSchedule.java

```diff
@@ -0,0 +1,559 @@
+                .map(FragmentsEdge::getTarget)
+                // filter stages that depend on completed stage only
+                .filter(dependentFragmentId -> fragmentDependency.inDegreeOf(dependentFragmentId) == 1)
+                .map(stagesByFragmentId::get)
+                .forEach(this::selectForExecution);
+        fragmentDependency.removeVertex(fragmentId);
+        fragmentTopology.removeVertex(fragmentId);
+        activeStages.remove(stage);
+    }
+
+    private void unblockStagesWithFullOutputBuffer()
+    {
+        // find stages that are blocked on full task output buffer
+        Set<PlanFragmentId> blockedFragments = activeStages.stream()
+                .filter(StageExecution::isAnyTaskBlocked)
+                .map(stage -> stage.getFragment().getId())
+                .collect(toImmutableSet());
+        // start immediate downstream stages so that data can be consumed
+        blockedFragments.stream()
+                .flatMap(fragmentId -> fragmentTopology.outgoingEdgesOf(fragmentId).stream())
+                .map(FragmentsEdge::getTarget)
+                .map(stagesByFragmentId::get)
+                .forEach(this::selectForExecution);
+    }
+
+    private void selectForExecution(StageExecution stage)
+    {
+        if (fragmentDependency.outDegreeOf(stage.getFragment().getId()) > 0) {
+            // if there are any dependent stages then reschedule when stage is completed
+            stage.addStateChangeListener(state -> {
+                if (isStageCompleted(stage)) {
+                    notifyReschedule();
+                }
+            });
+        }
+        activeStages.add(stage);
+    }
+
+    private void notifyReschedule()
+    {
+        SettableFuture<Void> rescheduleFuture;
+        synchronized (this) {
+            rescheduleFuture = this.rescheduleFuture;
+            this.rescheduleFuture = SettableFuture.create();
+        }
+        // notify listeners outside of the critical section
+        rescheduleFuture.set(null);
+    }
+
+    private boolean isStageCompleted(StageExecution stage)
+    {
+        State state = stage.getState();
+        return state == SCHEDULED || state == RUNNING || state == FLUSHING || state.isDone();
+    }
+
+    private Set<PlanFragmentId> extractDependenciesAndReturnNonLazyFragments(Collection<PlanFragment> fragments)
+    {
+        // Build a graph where the plan fragments are vertexes and the edges represent
+        // a before -> after relationship. Destination fragment should be started only
+        // when source fragment is completed. For example, a join hash build has an edge
+        // to the join probe.
+        fragments.forEach(fragment -> {
+            fragmentDependency.addVertex(fragment.getId());
+            fragmentTopology.addVertex(fragment.getId());
+        });
+
+        Visitor visitor = new Visitor(fragments);
+        fragments.forEach(fragment -> visitor.processFragment(fragment.getId()));
```

Comment:
nit: maybe put the loop inside visitor adding `Visitor.processAllFragments()` method.